lib/nat/interface.go

```diff
@@ -19,11 +19,19 @@ const (
 	UDP Protocol = "UDP"
 )
 
+type IPVersion int8
+
+const (
+	DualStack = iota
+	IPv4Only
+	IPv6Only
+)
+
 type Device interface {
 	ID() string
 	GetLocalIPv4Address() net.IP
 	AddPortMapping(ctx context.Context, protocol Protocol, internalPort, externalPort int, description string, duration time.Duration) (int, error)
-	AddPinhole(ctx context.Context, protocol Protocol, port int, duration time.Duration) ([]net.IP, error)
+	AddPinhole(ctx context.Context, protocol Protocol, addr Address, duration time.Duration) ([]net.IP, error)
```

Comment:
Why are we passing the address only for AddPinhole, not for AddPortMapping? Doesn't the same problem exist there?

Reply:
For IPv4, with the assumption that there really is only one IPv4 per interface the problem of using an interface that doesn't have the IP we're listening on should be solved by the `validGateway` check. 

Given that that's probably a valid assumption on most devices and more importantly, that the old code for getting *a* local IPv4 is still there as a fallback on OSes where we don't have permission to list all IPs belonging to an interface, I'd leave the IPv4 code as it currently is for now.

Reply:
> +1 from me to keep the scope of this PR limited.

Very much agreed.  
I have long lost track of what's happening here, but when I peeked in I got some strong bikeshedding vibes. Maybe that's my missing context. And it seems like @maxi0604 doesn't mind so far. Nevertheless please try to focus to get something good enough to add "IPv6 UPnP support", not fix all the problems related to connectivity in a future prove (beyond IPv6) way. Also smaller PRs are always better, easier to review and makes it easier to disect issues after merge if they come up.  
tl;dr: Thanks everyone for the work and enthusiasm here - please try to limit the scope of this PR/prevent feature-creep, there's always the possibility of another PR :)

Reply:
Sure, I don't want to extend the scope either.  It's just a rather complex affair with all possible combinations and networking being what it is.  So whatever comes to mind while trying to grok this code (new and old parts), I tend to ask questions and report findings.  I too hope that @maxi0604 doesn't mind my pickiness about coding style, and I tried to always give constructive hints including code suggestions.  I really like Syncthing's overall rather clean and tidy code base, thus applying a high standard while reviewing new code.  Small stylistic issues tend to distract me there, so why not just report them right away.  Again, hoping that it's perceived more as useful tips, not trying to be pedantic and unnerving :-)

Reply:
So far, it seems to me that the majority of work has been about actual issues that one discovers while writing the code and decision making not only for this implementation but also future implementations of protocols like this. I'm also happy to be reminded of coding style and to have the code reviewed for any issues, especially with so many "moving parts" involved. 

That said, I think that the PR has now reached a point where the IPv6 code works and is only triggered when IPv6 is actually present so if no more issues are discovered I'd be happy to keep structure of the code like it currently is and leave the limitations of the IPv4 code like they are.